test/e2e/upgrades/kube_proxy_migration.go

```diff
@@ -0,0 +1,222 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrades
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/api/core/v1"
+	extensions "k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	defaultTestTimeout   = time.Duration(5 * time.Minute)
+	clusterAddonLabelKey = "k8s-app"
+	clusterComponentKey  = "component"
+	kubeProxyLabelName   = "kube-proxy"
+)
+
+// KubeProxyUpgradeTest tests kube-proxy static pods -> DaemonSet upgrade path.
+type KubeProxyUpgradeTest struct {
+}
+
+func (KubeProxyUpgradeTest) Name() string { return "kube-proxy-upgrade" }
+
+// Setup does nothing.
+func (t *KubeProxyUpgradeTest) Setup(f *framework.Framework) {
+}
+
+// Test validates if kube-proxy is migrated from static pods to DaemonSet.
+func (t *KubeProxyUpgradeTest) Test(f *framework.Framework, done <-chan struct{}, upgrade UpgradeType) {
+	c := f.ClientSet
+
+	By("Waiting for kube-proxy static pods running and ready")
+	Expect(waitForKubeProxyStaticPodsRunning(c)).NotTo(HaveOccurred())
```

Comment:
This seems to make API calls. This will likely be very flaky as the master could be down during this. If you need to check that the static pods are running before an upgrade, you should do that in `Setup`.

Testing during an upgrade should be reserved for things expected to work during an upgrade. e.g. a replicated service with PDBs remains reachable.

Reply:
Thanks, yes we should certainly do that in `Setup`. Done.